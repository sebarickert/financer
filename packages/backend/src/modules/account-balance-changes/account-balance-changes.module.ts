import { Module } from '@nestjs/common';

import { DatabaseModule } from '../../database/prisma.module';

import { AccountBalanceChangesService } from './account-balance-changes.service';

@Module({
  imports: [DatabaseModule],
  providers: [AccountBalanceChangesService],
  exports: [AccountBalanceChangesService],
})
export class AccountBalanceChangesModule {}
