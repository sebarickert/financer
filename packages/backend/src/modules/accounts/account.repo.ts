import { Injectable } from '@nestjs/common';
import { Account, Prisma, PrismaPromise } from '@prisma/client';

import { PrismaService } from '@/database/prisma.service';

@Injectable()
export class AccountRepo {
  constructor(private readonly prisma: PrismaService) {}

  async findOne(
    accountWhereUniqueInput: Prisma.AccountWhereUniqueInput,
  ): Promise<Account | null> {
    return this.prisma.account.findUnique({
      where: accountWhereUniqueInput,
    });
  }

  async count(where: Prisma.AccountWhereInput): Promise<number> {
    return this.prisma.account.count({
      where,
    });
  }

  async findMany(params: {
    skip?: number;
    take?: number;
    cursor?: Prisma.AccountWhereUniqueInput;
    where?: Prisma.AccountWhereInput;
    orderBy?: Prisma.AccountOrderByWithRelationInput;
  }): Promise<Account[]> {
    const { skip, take, cursor, where, orderBy } = params;
    return this.prisma.account.findMany({
      skip,
      take,
      cursor,
      where,
      orderBy,
    });
  }

  async getCount(params: {
    skip?: number;
    take?: number;
    cursor?: Prisma.AccountWhereUniqueInput;
    where?: Prisma.AccountWhereInput;
    orderBy?: Prisma.AccountOrderByWithRelationInput;
  }): Promise<number> {
    const { skip, take, cursor, where, orderBy } = params;
    return this.prisma.account.count({
      skip,
      take,
      cursor,
      where,
      orderBy,
    });
  }

  async create(data: Prisma.AccountUncheckedCreateInput): Promise<Account> {
    const {
      id,
      userId,
      name,
      type,
      balance,
      createdAt,
      updatedAt,
      isDeleted,
      currentDateBalance,
    } = data;
    return this.prisma.account.create({
      data: {
        id,
        userId,
        name,
        type,
        balance,
        createdAt,
        updatedAt,
        isDeleted,
        currentDateBalance,
      },
    });
  }

  createMany(
    data: Prisma.AccountUncheckedCreateInput[],
  ): PrismaPromise<Prisma.BatchPayload> {
    return this.prisma.account.createMany({
      data,
    });
  }

  async update(params: {
    where: Prisma.AccountWhereUniqueInput;
    data: Prisma.AccountUpdateInput;
  }): Promise<Account> {
    const { where, data } = params;
    const {
      id,
      userId,
      name,
      type,
      balance,
      createdAt,
      updatedAt,
      isDeleted,
      currentDateBalance,
    } = data as Prisma.AccountUncheckedUpdateInput;
    return this.prisma.account.update({
      data: {
        id,
        userId,
        name,
        type,
        balance,
        createdAt,
        updatedAt,
        isDeleted,
        currentDateBalance,
      },
      where,
    });
  }

  async delete(where: Prisma.AccountWhereUniqueInput): Promise<Account> {
    return this.prisma.account.delete({
      where,
    });
  }

  deleteMany(where: Prisma.AccountWhereInput) {
    return this.prisma.account.deleteMany({
      where,
    });
  }
}
