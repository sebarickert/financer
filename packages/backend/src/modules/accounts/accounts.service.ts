import {
  forwardRef,
  Inject,
  Injectable,
  NotFoundException,
} from '@nestjs/common';
import { Account, AccountType } from '@prisma/client';

import { AccountRepo } from '../../database/repos/account.repo';
import { PaginationDto } from '../../types/pagination.dto';
import { sumArrayItems } from '../../utils/arrays';
import { AccountBalanceChangesService } from '../account-balance-changes/account-balance-changes.service';
import { TransactionsService } from '../transactions/transactions.service';

import { AccountBalanceHistoryDto } from './dto/account-balance-history.dto';
import { AccountDto } from './dto/account.dto';
import { CreateAccountDto } from './dto/create-account.dto';
import { UpdateAccountDto } from './dto/update-account.dto';

@Injectable()
export class AccountsService {
  constructor(
    private readonly accountRepo: AccountRepo,
    @Inject(forwardRef(() => TransactionsService))
    private transactionsService: TransactionsService,
    private accountBalanceChangeService: AccountBalanceChangesService,
  ) {}

  async create(
    userId: string,
    createAccountDto: CreateAccountDto,
  ): Promise<Account> {
    return this.accountRepo.create({ ...createAccountDto, userId });
  }

  async createMany(
    createAccountDto: CreateAccountDto[],
    userId: string,
  ): Promise<void> {
    await this.accountRepo.createMany(
      createAccountDto.map((account) => ({ ...account, userId })),
    );
  }

  async findOne(userId: string, id: string): Promise<Account> {
    const account = await this.accountRepo.findOne({ id, userId });

    if (!account) {
      throw new NotFoundException('Account not found.');
    }

    return account;
  }

  async findAllByUser(
    userId: string,
    accountTypes?: AccountType[],
    limit = 20,
    page = 1,
  ): Promise<PaginationDto<AccountDto[]>> {
    const skip = page * limit - limit;

    const accounts = await this.accountRepo.findMany({
      where: { userId, isDeleted: false, type: { in: accountTypes } },
      take: limit,
      skip,
    });

    const totalCount = await this.accountRepo.getCount({
      where: { userId, isDeleted: false, type: { in: accountTypes } },
    });

    const lastPage = page ? Math.ceil(totalCount / limit) : 1;

    return {
      data: accounts,
      currentPage: page ?? 1,
      limit: page ? limit : totalCount,
      totalPageCount: lastPage,
      hasNextPage: (page ?? 1) < lastPage,
      hasPreviousPage: (page ?? 1) > 1,
      totalRowCount: totalCount,
    };
  }

  async findAllIncludeDeletedByUser(userId: string): Promise<Account[]> {
    return this.accountRepo.findMany({ where: { userId } });
  }

  async update(
    userId: string,
    id: string,
    updateAccountDto: UpdateAccountDto,
  ): Promise<Account> {
    const accountBeforeChange = await this.findOne(userId, id);

    if (
      updateAccountDto.balance &&
      updateAccountDto.balance !== accountBeforeChange.balance
    ) {
      const balanceChangeAmount =
        updateAccountDto.balance - accountBeforeChange.balance;
      await this.accountBalanceChangeService.create({
        accountId: id.toString(),
        userId: userId.toString(),
        amount: balanceChangeAmount,
        date: new Date(),
      });
    }

    return this.accountRepo.update({
      where: { userId, id },
      data: updateAccountDto,
    });
  }

  async updateBalance(
    userId: string,
    id: string,
    amount: number,
  ): Promise<Account> {
    await this.findOne(userId, id);

    return this.accountRepo.update({
      where: { userId, id },
      data: { balance: { increment: amount } },
    });
  }

  async remove(id: string, userId: string) {
    await this.findOne(userId, id);
    await this.accountRepo.update({
      where: { id, userId },
      data: { isDeleted: true },
    });
  }

  async removeAllByUser(userId: string) {
    await this.accountRepo.deleteMany({ userId });
  }

  async getAccountBalanceHistory(
    userId: string,
    accountId: string,
  ): Promise<AccountBalanceHistoryDto[]> {
    const account = await this.findOne(userId, accountId);

    const accountBalanceChanges = (
      await this.accountBalanceChangeService.findAllByUserAndAccount(
        userId.toString(),
        accountId.toString(),
      )
    ).map(({ amount, date }) => ({ amount, date }));

    const accountTransactions = (
      await this.transactionsService.findAllByUser(
        userId,
        null,
        undefined,
        undefined,
        undefined,
        undefined,
        accountId,
      )
    ).data.map(({ amount, date, toAccount }) => ({
      date,
      amount: accountId === toAccount.toString() ? amount : -amount,
    }));

    const allBalanceChanges = accountBalanceChanges.concat(accountTransactions);

    const summarizedBalanceChanges = allBalanceChanges
      .reduce(
        (previous, { date }) =>
          previous.some((d) => d.getTime() === date.getTime())
            ? previous
            : previous.concat(date),
        [] as Date[],
      )
      .sort((a, b) => b.getTime() - a.getTime())
      .map((date) => {
        const balanceChangeAmounts = allBalanceChanges
          .filter(({ date: itemDate }) => itemDate.getTime() === date.getTime())
          .map(({ amount }) => amount);

        return {
          date,
          amount: sumArrayItems(balanceChangeAmounts),
        };
      })
      .reduce((previous, { date, amount }) => {
        const previousBalance = previous.at(-1)?.balance ?? account.balance;

        const previousAmount = previous.at(-1)?.amount ?? 0;

        const newBalance = previousBalance - previousAmount;

        return previous.concat({
          date,
          amount,
          balance: newBalance,
        });
      }, [] as AccountBalanceHistoryDto[]);

    return summarizedBalanceChanges;
  }

  private getAccountTypeFilter(accountTypes?: AccountType[]) {
    if (!accountTypes?.length) return {};

    return {
      type: {
        $in: accountTypes,
      },
    };
  }
}
