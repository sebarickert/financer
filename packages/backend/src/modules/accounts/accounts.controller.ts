import { AccountType } from '@local/types';
import {
  Controller,
  Get,
  Post,
  Body,
  Patch,
  Param,
  Delete,
  Query,
  ParseArrayPipe,
} from '@nestjs/common';
import {
  ApiBody,
  ApiExtraModels,
  ApiOkResponse,
  ApiParam,
  ApiQuery,
  ApiTags,
} from '@silte/nestjs-swagger';

import { ObjectId } from '../../types/objectId';
import { ApiPaginatedDto } from '../../utils/pagination.decorator';
import { ValidateEntityIdOld } from '../../utils/validate-entity-id.pipe';
import { LoggedIn } from '../auth/decorators/loggedIn.decorators';
import { UserIdOld } from '../users/users.decorators';

import { AccountsService } from './accounts.service';
import { AccountBalanceHistoryDto } from './dto/account-balance-history.dto';
import { AccountDto } from './dto/account.dto';
import { CreateAccountDto } from './dto/create-account.dto';
import { UpdateAccountDto } from './dto/update-account.dto';

@Controller('api/accounts')
@ApiTags('Accounts')
@ApiExtraModels(AccountDto)
@LoggedIn()
export class AccountsController {
  constructor(private readonly accountsService: AccountsService) {}

  @Post()
  @ApiBody({ type: CreateAccountDto })
  @ApiOkResponse({ schema: { properties: { payload: { type: 'string' } } } })
  async create(
    @UserIdOld() userId: ObjectId,
    @Body() createAccountDto: CreateAccountDto,
  ) {
    return this.accountsService.create(userId, createAccountDto);
  }

  @Get()
  @ApiPaginatedDto(AccountDto)
  @ApiQuery({
    name: 'limit',
    required: false,
  })
  @ApiQuery({
    name: 'page',
    required: false,
  })
  @ApiQuery({
    name: 'accountTypes',
    required: false,
  })
  async findAllByUser(
    @UserIdOld() userId: string,
    @Query('limit') limit?: number,
    @Query('page') page?: number,
    @Query(
      'accountTypes',
      new ParseArrayPipe({ separator: '|', optional: true }),
    )
    accountTypes?: AccountType[],
  ) {
    return this.accountsService.findAllByUser(
      userId,
      accountTypes,
      limit || undefined,
      page || undefined,
    );
  }

  @Get(':id')
  @ApiOkResponse({
    type: AccountDto,
    description: 'Return account by id',
  })
  @ApiParam({
    name: 'id',
    type: String,
  })
  async findOneById(
    @UserIdOld() userId: ObjectId,
    @Param('id', ValidateEntityIdOld) id: ObjectId,
  ) {
    return this.accountsService.findOne(userId, id);
  }

  @Patch(':id')
  @ApiBody({ type: UpdateAccountDto })
  @ApiOkResponse({ type: AccountDto })
  @ApiParam({
    name: 'id',
    type: String,
  })
  async update(
    @UserIdOld() userId: ObjectId,
    @Param('id', ValidateEntityIdOld) id: ObjectId,
    @Body() updateAccountDto: UpdateAccountDto,
  ) {
    return this.accountsService.update(userId, id, updateAccountDto);
  }

  @Delete(':id')
  @ApiParam({
    name: 'id',
    type: String,
  })
  remove(@UserIdOld() userId: ObjectId, @Param('id') id: ObjectId) {
    return this.accountsService.remove(id, userId);
  }

  @Get(':id/balance-history')
  @ApiOkResponse({
    type: AccountBalanceHistoryDto,
    isArray: true,
    description: 'Return account balance history by id',
  })
  @ApiParam({
    name: 'id',
    type: String,
  })
  async getAccountBalanceHistory(
    @UserIdOld() userId: ObjectId,
    @Param('id', ValidateEntityIdOld) id: ObjectId,
  ) {
    return this.accountsService.getAccountBalanceHistory(userId, id);
  }
}
