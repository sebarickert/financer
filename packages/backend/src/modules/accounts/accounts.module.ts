import { forwardRef, Module } from '@nestjs/common';

import { DatabaseModule } from '../../database/database.module';
import { AccountBalanceChangesModule } from '../account-balance-changes/account-balance-changes.module';
import { TransactionsModule } from '../transactions/transactions.module';

import { AccountsController } from './accounts.controller';
import { AccountsService } from './accounts.service';

@Module({
  imports: [
    DatabaseModule,
    AccountBalanceChangesModule,
    forwardRef(() => TransactionsModule),
  ],
  controllers: [AccountsController],
  providers: [AccountsService],
  exports: [AccountsService],
})
export class AccountsModule {}
