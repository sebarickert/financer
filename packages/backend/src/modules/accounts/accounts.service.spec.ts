import { Test, TestingModule } from '@nestjs/testing';
import { Decimal } from '@prisma/client/runtime/library';

import { AccountsService } from './accounts.service';

import { AccountBalanceChangesService } from '@/account-balance-changes/account-balance-changes.service';
import { DUMMY_TEST_USER } from '@/config/mockAuthenticationMiddleware';
import { AccountBalanceChangeRepo } from '@/modules/account-balance-changes/account-balance-change.repo';
import { AccountRepo } from '@/modules/accounts/account.repo';
import {
  accountsRepoFindAllMockData,
  accountsRepoFindById,
} from '@/database/repos/mocks/account-repo-mock';
import { transactionsRepoFindAllByAccountIdMockData } from '@/database/repos/mocks/transactions-repo-mock';
import { TransactionRepo } from '@/modules/transactions/transaction.repo';
import { createMockServiceProvider } from '@/test/create-mock-service-provider';
import { TransactionCategoriesService } from '@/transaction-categories/transaction-categories.service';
import { TransactionCategoryMappingsService } from '@/transaction-category-mappings/transaction-category-mappings.service';
import { TransactionsService } from '@/transactions/transactions.service';

describe('AccountsService', () => {
  let service: AccountsService;
  let accountRepo: jest.Mocked<AccountRepo>;
  let accountBalanceChangeRepo: jest.Mocked<AccountBalanceChangeRepo>;
  let transactionRepo: jest.Mocked<TransactionRepo>;
  let accountBalanceChangesService: AccountBalanceChangesService;
  let transactionsService: TransactionsService;

  beforeEach(async () => {
    jest.clearAllMocks();

    const module: TestingModule = await Test.createTestingModule({
      providers: [
        AccountsService,
        AccountBalanceChangesService,
        TransactionsService,
        createMockServiceProvider(TransactionCategoriesService),
        createMockServiceProvider(TransactionCategoryMappingsService),
        createMockServiceProvider(AccountRepo),
        createMockServiceProvider(AccountBalanceChangeRepo),
        createMockServiceProvider(TransactionRepo),
      ],
    }).compile();

    service = module.get<AccountsService>(AccountsService);
    accountRepo = module.get<jest.Mocked<AccountRepo>>(AccountRepo);
    accountBalanceChangeRepo = module.get<
      jest.Mocked<AccountBalanceChangeRepo>
    >(AccountBalanceChangeRepo);
    transactionRepo = module.get<jest.Mocked<TransactionRepo>>(TransactionRepo);
    accountBalanceChangesService = module.get<AccountBalanceChangesService>(
      AccountBalanceChangesService,
    );
    transactionsService = module.get<TransactionsService>(TransactionsService);
  });

  it('should return an array of accounts from findAllByUser', async () => {
    jest
      .spyOn(accountRepo, 'findMany')
      .mockResolvedValueOnce(accountsRepoFindAllMockData);

    jest
      .spyOn(service, 'getCurrentDateAccountBalance')
      .mockResolvedValue(new Decimal(0));

    jest.spyOn(accountRepo, 'getCount').mockResolvedValueOnce(9);

    const accounts = await service.findAllByUser(DUMMY_TEST_USER.id);

    expect(accountRepo.findMany).toHaveBeenCalledTimes(1);
    expect(accountRepo.findMany).toHaveBeenCalledWith({
      where: {
        isDeleted: false,
        userId: '61460d7354ea082ad0256749',
      },
    });

    expect(accounts).toMatchSnapshot();
  });

  it('should return an account from findOneWithCurrentBalance', async () => {
    const id = '61460d8554ea082ad0256759';

    jest
      .spyOn(accountRepo, 'findOne')
      .mockResolvedValueOnce(accountsRepoFindById[id]);

    jest
      .spyOn(accountBalanceChangesService, 'findAllByUserAndAccount')
      .mockResolvedValueOnce([]);

    jest
      .spyOn(transactionsService, 'findTransactionSummariesByUserAccount')
      .mockResolvedValueOnce([]);

    const account = await service.findOneWithCurrentBalance(
      DUMMY_TEST_USER.id,
      id,
    );

    expect(accountRepo.findOne).toHaveBeenCalledTimes(1);
    expect(accountRepo.findOne).toHaveBeenCalledWith({
      id: '61460d8554ea082ad0256759',
      userId: '61460d7354ea082ad0256749',
    });

    expect(account).toMatchSnapshot();
  });

  it('should return an array of account balance history from getAccountBalanceHistory', async () => {
    const id = '61460d8554ea082ad0256759';

    jest.spyOn(transactionsService, 'findAllByUser').mockResolvedValueOnce([]);

    jest.spyOn(accountBalanceChangeRepo, 'findMany').mockResolvedValueOnce([]);

    jest
      .spyOn(transactionRepo, 'findMany')
      .mockResolvedValueOnce(transactionsRepoFindAllByAccountIdMockData[id]);

    const accountBalanceHistory = await service.getAccountBalanceHistory(
      DUMMY_TEST_USER.id,
      accountsRepoFindById[id],
    );

    expect(accountBalanceChangeRepo.findMany).toHaveBeenCalledTimes(1);
    expect(accountBalanceChangeRepo.findMany).toHaveBeenCalledWith({
      where: {
        accountId: '61460d8554ea082ad0256759',
        userId: '61460d7354ea082ad0256749',
      },
    });

    expect(transactionRepo.findMany).toHaveBeenCalledTimes(1);
    expect(transactionRepo.findMany).toHaveBeenCalledWith({
      select: {
        amount: true,
        date: true,
        fromAccount: true,
        id: true,
        toAccount: true,
      },
      where: {
        OR: [
          {
            toAccount: '61460d8554ea082ad0256759',
          },
          {
            fromAccount: '61460d8554ea082ad0256759',
          },
        ],
        date: undefined,
        userId: '61460d7354ea082ad0256749',
      },
    });

    expect(accountBalanceHistory).toMatchSnapshot();
  });
});
