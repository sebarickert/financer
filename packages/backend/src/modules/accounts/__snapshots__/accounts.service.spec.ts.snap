// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AccountsService should return an account from findOne 1`] = `
AccountDto {
  "balance": "5728.67",
  "createdAt": 2024-05-10T10:21:18.333Z,
  "id": "61460d8554ea082ad0256759",
  "isDeleted": false,
  "name": "Saving account 1",
  "type": "SAVINGS",
  "updatedAt": 2024-05-10T10:21:18.333Z,
  "userId": "61460d7354ea082ad0256749",
}
`;

exports[`AccountsService should return an array of account balance history from getAccountBalanceHistory 1`] = `
[
  AccountBalanceHistoryDto {
    "amount": "318.22",
    "balance": "118.22",
    "date": 2021-02-21T19:23:00.000Z,
  },
  AccountBalanceHistoryDto {
    "amount": "459.72",
    "balance": "577.94",
    "date": 2021-02-23T13:28:00.000Z,
  },
  AccountBalanceHistoryDto {
    "amount": "1704.16",
    "balance": "2282.1",
    "date": 2021-02-28T01:30:00.000Z,
  },
  AccountBalanceHistoryDto {
    "amount": "1901.59",
    "balance": "4183.69",
    "date": 2021-03-01T03:31:00.000Z,
  },
  AccountBalanceHistoryDto {
    "amount": "-118.27",
    "balance": "4065.42",
    "date": 2021-03-01T09:31:00.000Z,
  },
  AccountBalanceHistoryDto {
    "amount": "-109.53",
    "balance": "3955.89",
    "date": 2021-03-05T13:18:00.000Z,
  },
  AccountBalanceHistoryDto {
    "amount": "-1279.27",
    "balance": "2676.62",
    "date": 2021-03-09T09:08:00.000Z,
  },
  AccountBalanceHistoryDto {
    "amount": "-1368.87",
    "balance": "1307.75",
    "date": 2021-03-21T11:16:00.000Z,
  },
  AccountBalanceHistoryDto {
    "amount": "1590.98",
    "balance": "2898.73",
    "date": 2021-03-23T07:54:00.000Z,
  },
  AccountBalanceHistoryDto {
    "amount": "1494.56",
    "balance": "4393.29",
    "date": 2021-03-27T04:19:00.000Z,
  },
  AccountBalanceHistoryDto {
    "amount": "-1403.68",
    "balance": "2989.61",
    "date": 2021-04-09T17:38:00.000Z,
  },
  AccountBalanceHistoryDto {
    "amount": "86.6",
    "balance": "3076.21",
    "date": 2021-04-16T06:52:00.000Z,
  },
  AccountBalanceHistoryDto {
    "amount": "-663.32",
    "balance": "2412.89",
    "date": 2021-05-18T20:51:00.000Z,
  },
  AccountBalanceHistoryDto {
    "amount": "675",
    "balance": "3087.89",
    "date": 2021-05-25T03:50:00.000Z,
  },
  AccountBalanceHistoryDto {
    "amount": "1925.73",
    "balance": "5013.62",
    "date": 2021-05-30T08:06:00.000Z,
  },
  AccountBalanceHistoryDto {
    "amount": "419.53",
    "balance": "5433.15",
    "date": 2021-06-04T09:01:00.000Z,
  },
  AccountBalanceHistoryDto {
    "amount": "1292.28",
    "balance": "6725.43",
    "date": 2021-06-08T03:28:00.000Z,
  },
  AccountBalanceHistoryDto {
    "amount": "1299.2",
    "balance": "8024.63",
    "date": 2021-06-17T09:39:00.000Z,
  },
  AccountBalanceHistoryDto {
    "amount": "-490.7",
    "balance": "7533.93",
    "date": 2021-06-20T06:36:00.000Z,
  },
  AccountBalanceHistoryDto {
    "amount": "-1865",
    "balance": "5668.93",
    "date": 2021-06-25T10:02:00.000Z,
  },
  AccountBalanceHistoryDto {
    "amount": "961.29",
    "balance": "6630.22",
    "date": 2021-07-13T11:55:00.000Z,
  },
  AccountBalanceHistoryDto {
    "amount": "-129.94",
    "balance": "6500.28",
    "date": 2021-07-16T04:54:00.000Z,
  },
  AccountBalanceHistoryDto {
    "amount": "1472.83",
    "balance": "7973.11",
    "date": 2021-07-20T02:50:00.000Z,
  },
  AccountBalanceHistoryDto {
    "amount": "723.92",
    "balance": "8697.03",
    "date": 2021-07-24T13:04:00.000Z,
  },
  AccountBalanceHistoryDto {
    "amount": "-1715.66",
    "balance": "6981.37",
    "date": 2021-07-25T23:53:00.000Z,
  },
  AccountBalanceHistoryDto {
    "amount": "1975.42",
    "balance": "8956.79",
    "date": 2021-07-28T03:35:00.000Z,
  },
  AccountBalanceHistoryDto {
    "amount": "-131.18",
    "balance": "8825.61",
    "date": 2021-07-29T17:30:00.000Z,
  },
  AccountBalanceHistoryDto {
    "amount": "1799.66",
    "balance": "10625.27",
    "date": 2021-07-31T20:26:00.000Z,
  },
  AccountBalanceHistoryDto {
    "amount": "-1887.67",
    "balance": "8737.6",
    "date": 2021-08-06T07:37:00.000Z,
  },
  AccountBalanceHistoryDto {
    "amount": "1358.23",
    "balance": "10095.83",
    "date": 2021-08-14T12:45:00.000Z,
  },
  AccountBalanceHistoryDto {
    "amount": "1700.08",
    "balance": "11795.91",
    "date": 2021-08-22T13:14:00.000Z,
  },
  AccountBalanceHistoryDto {
    "amount": "441.11",
    "balance": "12237.02",
    "date": 2021-08-24T11:03:00.000Z,
  },
  AccountBalanceHistoryDto {
    "amount": "348.62",
    "balance": "12585.64",
    "date": 2021-08-25T12:24:00.000Z,
  },
  AccountBalanceHistoryDto {
    "amount": "-858.43",
    "balance": "11727.21",
    "date": 2021-08-27T22:37:00.000Z,
  },
  AccountBalanceHistoryDto {
    "amount": "-1649.03",
    "balance": "10078.18",
    "date": 2021-08-29T10:47:00.000Z,
  },
  AccountBalanceHistoryDto {
    "amount": "-1832.52",
    "balance": "8245.66",
    "date": 2021-08-31T08:33:00.000Z,
  },
  AccountBalanceHistoryDto {
    "amount": "1360.98",
    "balance": "9606.64",
    "date": 2021-09-05T12:53:00.000Z,
  },
  AccountBalanceHistoryDto {
    "amount": "-1740.36",
    "balance": "7866.28",
    "date": 2021-09-05T15:24:00.000Z,
  },
  AccountBalanceHistoryDto {
    "amount": "-849.95",
    "balance": "7016.33",
    "date": 2021-09-15T12:16:00.000Z,
  },
  AccountBalanceHistoryDto {
    "amount": "-127.84",
    "balance": "6888.49",
    "date": 2021-09-15T19:54:00.000Z,
  },
  AccountBalanceHistoryDto {
    "amount": "1605.25",
    "balance": "8493.74",
    "date": 2021-09-16T04:19:00.000Z,
  },
  AccountBalanceHistoryDto {
    "amount": "-1646.9",
    "balance": "6846.84",
    "date": 2021-09-17T17:57:00.000Z,
  },
  AccountBalanceHistoryDto {
    "amount": "1206.49",
    "balance": "8053.33",
    "date": 2021-09-22T05:52:00.000Z,
  },
  AccountBalanceHistoryDto {
    "amount": "-1178.6",
    "balance": "6874.73",
    "date": 2021-10-03T07:34:00.000Z,
  },
  AccountBalanceHistoryDto {
    "amount": "-729.58",
    "balance": "6145.15",
    "date": 2021-10-03T21:26:00.000Z,
  },
  AccountBalanceHistoryDto {
    "amount": "548.07",
    "balance": "6693.22",
    "date": 2021-10-11T14:51:00.000Z,
  },
  AccountBalanceHistoryDto {
    "amount": "447.51",
    "balance": "7140.73",
    "date": 2021-10-17T08:37:00.000Z,
  },
  AccountBalanceHistoryDto {
    "amount": "-1384.83",
    "balance": "5755.9",
    "date": 2021-10-23T00:56:00.000Z,
  },
  AccountBalanceHistoryDto {
    "amount": "1216.58",
    "balance": "6972.48",
    "date": 2021-10-24T17:57:00.000Z,
  },
  AccountBalanceHistoryDto {
    "amount": "-1817.43",
    "balance": "5155.05",
    "date": 2021-10-30T03:10:00.000Z,
  },
  AccountBalanceHistoryDto {
    "amount": "-656.84",
    "balance": "4498.21",
    "date": 2021-11-03T08:57:00.000Z,
  },
  AccountBalanceHistoryDto {
    "amount": "-1137.47",
    "balance": "3360.74",
    "date": 2021-11-05T02:31:00.000Z,
  },
  AccountBalanceHistoryDto {
    "amount": "1141.4",
    "balance": "4502.14",
    "date": 2021-11-06T00:38:00.000Z,
  },
  AccountBalanceHistoryDto {
    "amount": "469.16",
    "balance": "4971.3",
    "date": 2021-11-06T22:06:00.000Z,
  },
  AccountBalanceHistoryDto {
    "amount": "1666.29",
    "balance": "6637.59",
    "date": 2021-11-07T15:43:00.000Z,
  },
  AccountBalanceHistoryDto {
    "amount": "-879.34",
    "balance": "5758.25",
    "date": 2021-11-08T15:54:00.000Z,
  },
  AccountBalanceHistoryDto {
    "amount": "150.89",
    "balance": "5909.14",
    "date": 2021-11-15T14:28:00.000Z,
  },
  AccountBalanceHistoryDto {
    "amount": "-725.02",
    "balance": "5184.12",
    "date": 2021-11-30T09:45:00.000Z,
  },
  AccountBalanceHistoryDto {
    "amount": "-1563.22",
    "balance": "3620.9",
    "date": 2021-12-01T09:59:00.000Z,
  },
  AccountBalanceHistoryDto {
    "amount": "-852.27",
    "balance": "2768.63",
    "date": 2021-12-07T06:16:00.000Z,
  },
  AccountBalanceHistoryDto {
    "amount": "-1694.58",
    "balance": "1074.05",
    "date": 2021-12-10T01:22:00.000Z,
  },
  AccountBalanceHistoryDto {
    "amount": "970.41",
    "balance": "2044.46",
    "date": 2021-12-12T19:49:00.000Z,
  },
  AccountBalanceHistoryDto {
    "amount": "-873.08",
    "balance": "1171.38",
    "date": 2021-12-15T08:10:00.000Z,
  },
  AccountBalanceHistoryDto {
    "amount": "-306.23",
    "balance": "865.15",
    "date": 2021-12-15T16:07:00.000Z,
  },
  AccountBalanceHistoryDto {
    "amount": "-59.54",
    "balance": "805.61",
    "date": 2021-12-29T17:55:00.000Z,
  },
  AccountBalanceHistoryDto {
    "amount": "1786.09",
    "balance": "2591.7",
    "date": 2022-01-06T08:33:00.000Z,
  },
  AccountBalanceHistoryDto {
    "amount": "621.79",
    "balance": "3213.49",
    "date": 2022-01-07T14:37:00.000Z,
  },
  AccountBalanceHistoryDto {
    "amount": "1712.21",
    "balance": "4925.7",
    "date": 2022-01-18T23:33:00.000Z,
  },
  AccountBalanceHistoryDto {
    "amount": "1927.77",
    "balance": "6853.47",
    "date": 2022-01-20T06:13:00.000Z,
  },
  AccountBalanceHistoryDto {
    "amount": "-1124.8",
    "balance": "5728.67",
    "date": 2022-01-21T07:18:00.000Z,
  },
]
`;

exports[`AccountsService should return an array of accounts from findAllByUser 1`] = `
[
  AccountDto {
    "balance": "5728.67",
    "createdAt": 2024-05-10T10:21:18.333Z,
    "id": "61460d8554ea082ad0256759",
    "isDeleted": false,
    "name": "Saving account 1",
    "type": "SAVINGS",
    "updatedAt": 2024-05-10T10:21:18.333Z,
    "userId": "61460d7354ea082ad0256749",
  },
  AccountDto {
    "balance": "56323.6",
    "createdAt": 2024-05-10T10:21:18.333Z,
    "id": "61460d9454ea082ad0256762",
    "isDeleted": false,
    "name": "Saving account 2",
    "type": "SAVINGS",
    "updatedAt": 2024-05-10T10:21:18.333Z,
    "userId": "61460d7354ea082ad0256749",
  },
  AccountDto {
    "balance": "11201.26",
    "createdAt": 2024-05-10T10:21:18.333Z,
    "id": "61460da354ea082ad025676b",
    "isDeleted": false,
    "name": "Cash account",
    "type": "CASH",
    "updatedAt": 2024-05-10T10:21:18.333Z,
    "userId": "61460d7354ea082ad0256749",
  },
  AccountDto {
    "balance": "-2901.87",
    "createdAt": 2024-05-10T10:21:18.333Z,
    "id": "61460db554ea082ad0256774",
    "isDeleted": false,
    "name": "Investment account",
    "type": "INVESTMENT",
    "updatedAt": 2024-05-10T10:21:18.333Z,
    "userId": "61460d7354ea082ad0256749",
  },
  AccountDto {
    "balance": "3509.96",
    "createdAt": 2024-05-10T10:21:18.333Z,
    "id": "61460dd554ea082ad025677d",
    "isDeleted": false,
    "name": "Credit account",
    "type": "CREDIT",
    "updatedAt": 2024-05-10T10:21:18.333Z,
    "userId": "61460d7354ea082ad0256749",
  },
  AccountDto {
    "balance": "-6716.63",
    "createdAt": 2024-05-10T10:21:18.333Z,
    "id": "61460de154ea082ad0256786",
    "isDeleted": false,
    "name": "Loan account",
    "type": "LOAN",
    "updatedAt": 2024-05-10T10:21:18.333Z,
    "userId": "61460d7354ea082ad0256749",
  },
  AccountDto {
    "balance": "12400",
    "createdAt": 2024-05-10T10:22:18.423Z,
    "id": "663df55ad8ef53dcb2bc9347",
    "isDeleted": false,
    "name": "Big money",
    "type": "SAVINGS",
    "updatedAt": 2024-05-10T10:28:36.672Z,
    "userId": "61460d7354ea082ad0256749",
  },
  AccountDto {
    "balance": "21100",
    "createdAt": 2024-05-10T10:25:39.733Z,
    "id": "663df623d8ef53dcb2bc93c0",
    "isDeleted": false,
    "name": "Long-term SAVINGS",
    "type": "LONG_TERM_SAVINGS",
    "updatedAt": 2024-05-10T10:28:36.670Z,
    "userId": "61460d7354ea082ad0256749",
  },
  AccountDto {
    "balance": "14000",
    "createdAt": 2024-05-10T10:25:48.973Z,
    "id": "663df62cd8ef53dcb2bc93c4",
    "isDeleted": false,
    "name": "Pre-assigned CASH",
    "type": "PRE_ASSIGNED_CASH",
    "updatedAt": 2024-05-10T10:27:05.349Z,
    "userId": "61460d7354ea082ad0256749",
  },
]
`;
